index.html

```diff
@@ -2435,37 +2435,58 @@ <h2>Cryptographic Suites</h2>
 verification of the data.
         </li>
         <li>
-          The specification MUST define a <dfn>suite selection algorithm</dfn> that accepts a [=JSON
-          object=] |proof| and uses the properties described in [[[#proofs]]] to return either
-          failure or a [=VC cryptographic suite=] [[Infra]] [=struct=]. This algorithm SHOULD be
-          registered in [[[VC-SPECS]]]. A
-          <dfn class="export">VC cryptographic suite</dfn> [=struct=] has the following
-          [=struct/items=].
+The specification MUST define a <dfn>suite selection algorithm</dfn> that
+accepts a set of options ([=map=] |options|) and returns a
+[=cryptographic suite=] ([=struct=] |cryptosuite|). This algorithm SHOULD be
+registered in [[[VC-SPECS]]]. A cryptographic suite [=struct=] has the
+following [=struct/items=].
           <dl data-dfn-for="cryptographic suite">
             <dt><dfn>transform</dfn></dt>
-            <dd>A [=transformation algorithm=] that takes a [=JSON object=] as input and returns a
-            [=JSON object=].</dd>
+            <dd>
+A [=transformation algorithm=] that takes an [=unsecured data document=]
+([=JSON object=] |unsecuredDocument|) and transformation options
+([=struct=] |options|) as input and produces a
+<em>transformed data document</em> ([=JSON object=] |transformedDocument|),
+or an error, as output.
+            </dd>
             <dt><dfn>hash</dfn></dt>
-            <dd>A [=hashing algorithm=] that takes a [=JSON object=] as input and returns a [=byte
-            sequence=].</dd>
-            <dt><dfn>proof creation options</dfn></dt>
             <dd>
-              <p>A [=struct=] type that needs to be supplied to the [=cryptographic suite/serialize
-              proof=] operation.</p>
-              <p class="note">To sign data, cryptographic suites need to be able to use a private
-              key. This private key will often be implied by a field in the [=proof creation
-              options=], but it is often a better design to avoid passing the key directly. Passing
-              a key identifier or an asynchronous signing algorithm allows implementations to take
-              advantage of hardware security modules (HSMs) to get better security.</p>
+A [=hashing algorithm=] that takes a transformed data document
+([=JSON object=] |transformedDocument|) and canonical proof configuration
+([=struct=] |canonicalProofConfig|) as input and returns a [=byte sequence=],
+or an error, as output.
```

Comment:
I need to look more closely at this, but it seems like it might be slightly misaligned or overly prescriptive for use with selective disclosure cryptosuites that return other cryptosuite-specific outputs (e.g., https://w3c.github.io/vc-di-ecdsa/#base-proof-hashing-ecdsa-sd-2023 returns an object with multiple values, some hashes and others not, as key-value pairs).

I'm concerned we're trying too hard to pin down an interface here rather than providing core functions and a base, loose, architecture for extending cryptosuites to fit into. There is generally a transformation step, which always takes the unsecured data document as an input -- but the output can be specialized to the cryptosuite -- and, generally, a hashing step follows that receives this cryptosuite-specific output and generates its own cryptosuite-specific output, which is then followed by a proof generation/serialization step that finally outputs the data integrity proof object.

In short, there's a pipeline that is cryptosuite specific that always starts with an unsecured document and ends up with a data integrity proof at the end of it, but cryptosuites should be free to define their own intermediates. The fact that the data integrity core spec provides some useful primitives to use there does not mean that every cryptosuite needs to use them exactly as they are for "simple digital signatures". What is required, however, is unsecured document in, and data integrity proof out (for the entire pipeline).

I also think it's useful to keep the "proof configuration options" typing information we've added here, but we might need to relax some of the "pipeline intermediates" to prevent unnecessarily restricting innovation in cryptosuites in that space. It's also just not even aligned with current selective disclosure suites (slightly as mentioned).

Reply:
+1 to the idea of merging the "transform", "hash", and "serializeProof" algorithms for the purposes of this interface. https://github.com/w3c/vc-data-integrity/pull/226#discussion_r1419986552 listed a couple options for improving the flexibility here, but since the only use of this interface just calls the operations as a pipeline, merging them seems best.

Reply:
Ok, we had a conversation about this during the call yesterday and are taking @jyasskin and @dlongley's guidance and fixing up the interface to be simpler / higher level.

Reply:
Ok, reworked the entire PR, but only focused on the interface definition and the usage of the interface in Add Proof and Verify Proof. It was more painful than I was hoping it would be... but it might be closer to what folks want now. I still need to work on the other side of the implementation in the matching ECDSA PR.

We would still need to do alignments w/ the Proof Set/Chain stuff in a later PR.